systemd: fix non-systemd configuration support

kamalesh-babulal opened the pull request:
`systemd` support is configurable and is enabled by default, but can
be disabled too, by configuring with `--enable-systemd=no` option
and the programs linking `-lcgroup` complains about missing header
file `libcgroup/systemd.h`, in cases with `systemd` support disabled.
The reason is that `libcgroup/systemd.h` is included via `libcgroup.h`
independent of `systemd` configuration but only installed only if the
`systemd` support is enabled.

This patchset attempts to fix the warning by installing `libcgroup/systemd.h`
in both the `--enable-systemd=yes,no` configurations and also
introduce systemd function stubs, for the functions available to the
users, these versions of functions throw error about missing `systemd`
support, when gets called. It also fixes the `tools` to handle the
non-systemd stub function's return value.

Fixes: #413

drakenclimber commented on tests/ftests/systemd.py:
```diff
@@ -70,7 +71,14 @@ def remove_scope_slice_conf(config, _slice, scope, controller, config_fname=None
+
+    # Check if libcgroup is compiled with --enable-systemd
+    @staticmethod
+    def is_systemd_enabled(config):
```
nitpick - `config` isn't being used.  Do you have a plan for it in a later commit?

kamalesh-babulal replied:
@drakenclimber Thank you so much for reviewing.  `config` is not required, will remove it.

drakenclimber commented on tests/ftests/systemd.py:
```diff
@@ -70,7 +71,14 @@ def remove_scope_slice_conf(config, _slice, scope, controller, config_fname=None
+    # Check if libcgroup is compiled with --enable-systemd
+    @staticmethod
+    def is_systemd_enabled(config):
+        if "create_scope" in dir(Cgroup):
```
Nifty :)

drakenclimber commented on tests/ftests/078-sudo-cgcreate_systemd_scope.py:
```diff
@@ -7,9 +7,10 @@
 from process import Process
+from systemd import Systemd
 from libcgroup import Mode
+from cgroup import Cgroup
```
Do you need this import?  I don't see any diffs that require it

kamalesh-babulal replied:
Yeah, it is not required by this change. It was more of a cosmetic change, to have a reverse xmas tree style

drakenclimber commented on tests/ftests/079-sudo-cgcreate_default_systemd_scope.py:
```diff
@@ -7,9 +7,10 @@
 from process import Process
+from systemd import Systemd
 from libcgroup import Mode
+from cgroup import Cgroup
```
Ditto.  Is this needed?

kamalesh-babulal replied:
It's a cosmetic change, to have a reverse xmas tree style.

drakenclimber commented on tests/ftests/084-sudo-cgcreate_systemd_scope_pid.py:
```diff
@@ -7,9 +7,10 @@
 from process import Process
+from systemd import Systemd
 from libcgroup import Mode
+from cgroup import Cgroup
```
Ditto ditto :)

kamalesh-babulal replied:
It's a cosmetic change, to have a reverse xmas tree style.

drakenclimber reviewed (commented):
@kamalesh-babulal looks good to me. I had a few silly questions/comments, but the meat of the changes look spot on.  Thanks for the hard work

drakenclimber commented:
> All of the automated tests complains about `is_systemd_enabled()`:
> 
> ```
>  Test Results:
> 	Run Date:                          Dec 04 15:27:33
> 	Passed:                                  3 test(s)
> 	Skipped:                                15 test(s)
> 	Failed:                                 12 test(s)
> 		Test:				071-sudo-set_default_systemd_cgroup.py - type object 'Systemd' has no attribute 'is_systemd_enabled'
> 		Test:				067-sudo-systemd_cgexec-v1.py - type object 
...
```

Argh, my bad, @kamalesh-babulal .  I put the wrong number of commits in my rebase script and completely mangled the cherry-pick.  Disregard these errors :( :(

kamalesh-babulal commented:
**v3 changes**:
- Remove function argument from `Systemd.is_systemd_enabled()` and its call sites (suggested by @drakenclimber)

kamalesh-babulal commented:
**v4 Changes**:
- Removed `#ifde WITH_SYSTEMD` from python binding
- Introduced `cgroup_is_systemd_enabled()` in `systemd.c`
- Modified `ftests/systemd.py::is_systemd_enabled` to used C implementation `cgroup_is_systemd_enabled()`
- Exported `cgroup_is_systemd_enabled()`
